src/apps/midas/binance-smart-chain/midas.pool.token-fetcher.ts

```diff
@@ -0,0 +1,91 @@
+import { Inject, Logger } from '@nestjs/common';
+import { utils } from 'ethers';
+
+import { IAppToolkit, APP_TOOLKIT } from '~app-toolkit/app-toolkit.interface';
+import { Register } from '~app-toolkit/decorators';
+import { ContractType } from '~position/contract.interface';
+import { PositionFetcher } from '~position/position-fetcher.interface';
+import { AppTokenPosition } from '~position/position.interface';
+import { Network } from '~types/network.interface';
+
+import { MidasContractFactory } from '../contracts';
+import { MIDAS_DEFINITION } from '../midas.definition';
+
+const appId = MIDAS_DEFINITION.id;
+const groupId = MIDAS_DEFINITION.groups.pool.id;
+const network = Network.BINANCE_SMART_CHAIN_MAINNET;
+
+const poolDirectoryAddress = '0x295d7347606f4bd810c8296bb8d75d657001fcf7';
+const poolLensAddress = '0x7ef21f8bfa2356cc6fe43755e46e8e07d4ee8a47';
+
+@Register.TokenPositionFetcher({ appId, groupId, network })
+export class BinanceSmartChainMidasPoolTokenFetcher implements PositionFetcher<AppTokenPosition> {
+  constructor(
+    @Inject(APP_TOOLKIT) private readonly appToolkit: IAppToolkit,
+    @Inject(MidasContractFactory) private readonly midasContractFactory: MidasContractFactory,
+  ) {}
+
+  logger = new Logger(BinanceSmartChainMidasPoolTokenFetcher.name);
+
+  async getPositions() {
+    try {
+      const poolDirectoryContract = this.midasContractFactory.midasPoolDirectory({
+        address: poolDirectoryAddress,
+        network,
+      });
+      const poolLensContract = this.midasContractFactory.midasPoolLens({
+        address: poolLensAddress,
+        network,
+      });
+
+      const [poolIndexes, pools] = await poolDirectoryContract.callStatic.getActivePools();
+
+      if (!pools.length || !poolIndexes.length) {
+        return [];
+      }
+
+      const tokens = await Promise.all(
+        poolIndexes.map(async poolId => {
+          const { comptroller, name } = await poolDirectoryContract.callStatic.pools(Number(poolId));
+
+          let supply = 0;
+
+          const assets = await poolLensContract.callStatic.getPoolAssetsWithData(comptroller);
+
+          assets.map(value => {
+            if (value.totalSupply) {
+              supply += Number(utils.formatUnits(value.totalSupply, value.underlyingDecimals));
+            }
+          });
+
+          const token: AppTokenPosition = {
+            type: ContractType.APP_TOKEN,
+            supply,
+            address: comptroller,
+            network,
+            price: 0,
+            symbol: '',
+            decimals: 0,
+            tokens: [],
+            dataProps: {},
+            displayProps: {
+              label: name,
+              images: [],
+            },
+            appId,
+            groupId,
+            pricePerShare: 0,
```

Comment:
Having pricePerShare and Price = 0 will result in balances = 0

Reply:
@smartdevcom we need to fetch the pricePerShare from the asset (cToken). 

We can calculate the price per share by as mentioned in the compound docs:

<img width="943" alt="Screenshot 2023-01-09 at 14 40 57" src="https://user-images.githubusercontent.com/839848/211321431-7b183c2e-ba02-4bc7-8df4-1f12adea2efa.png">

I took the screenshot from here:
https://docs.compound.finance/v2/#protocol-math

```typescript
...
const oneCTokenInUnderlying = exchangeRateCurrent / Math.pow(10, mantissa);
console.log('1 cBAT can be redeemed for', oneCTokenInUnderlying, 'BAT');
```

`oneCTokenInUnderlying` is the `pricePerShare` mentioned here.

@wpoulin I'm not to sure what the difference between `pricePerShare` and `price` of an `AppToken` is. I read the [docs](https://studio.zapper.fi/docs/concepts/app-tokens#what-are-the-properties-of-an-app-token) but still confused. Won't it be the same as the `pricePerShare` in the midas/compound case and only relevant for `AppTokens` which don't have a `pricePerShare` model? Should we set it to the same value as `pricePerShare` or is either of the property enough?

Reply:
In 99.9% of the cases, the price is coming from the `underlyingToken.price * pricePerShare`
E.g cUSDC (compound supply position)

The underlying of cUSDC is USDC. The pricePerShare (ratio) is 
  ```
    const [rateRaw, mantissa] = await Promise.all([this.getExchangeRate(params), this.getExchangeRateMantissa(params)]);
    return Number(rateRaw) / 10 ** mantissa;
  ```

Reply:
Ah yes sure makes total sense. We will adapt the PR 👍